File: web/src/api/system.ts
import { request } from '@/utils/request'

// 与 internal/api/v1/handler.HealthStatus 对齐。
export interface DatabaseCheck {
  driver: string
  ok: boolean
  error?: string
}

export interface HealthStatus {
  status: 'ok' | 'degraded'
  version: string
  commit?: string
  buildTime?: string
  uptimeSeconds: number
  database: DatabaseCheck
}

export function fetchHealth() {
  return request<HealthStatus>({ url: '/health', method: 'GET', nodeId: null, silent: true })
}

/* ---------- 系统总览：字段对齐后端 handler.OverviewResult 与 internal/service/sysinfo ---------- */

export interface HostInfo {
  hostname: string
  os: string
  osName: string
  kernel: string
  arch: string
  uptimeSeconds: number
  bootTime?: string
  now: string
  timezone: string
}

export interface CPUInfo {
  model: string
  cores: number
  usagePercent: number
  load1: number
  load5: number
  load15: number
}

export interface MemoryInfo {
  total: number
  available: number
  used: number
  buffers: number
  cached: number
  swapTotal: number
  swapUsed: number
  usedPercent: number
}

export interface DiskInfo {
  device: string
  path: string
  fsType: string
  total: number
  used: number
  free: number
  usedPercent: number
}

export interface NetInterfaceInfo {
  name: string
  bytesRecv: number
  bytesSent: number
  packetsRecv: number
  packetsSent: number
}

export interface PanelInfo {
  version: string
  commit?: string
  buildTime?: string
  goVersion: string
  pid: number
  startedAt: string
  uptimeSeconds: number
  goroutines: number
  heapAllocBytes: number
  rssBytes: number
  numCpu: number
  dataDir?: string
}

export interface DatabaseStat extends DatabaseCheck {
  openConnections: number
  inUse: number
  idle: number
}

export interface SystemOverview {
  host: HostInfo
  cpu: CPUInfo
  memory: MemoryInfo
  disks: DiskInfo[]
  network: NetInterfaceInfo[]
  panel: PanelInfo
  database: DatabaseStat
  /** 为 false 时 CPU/内存/磁盘/网络无法采集（非 Linux 主机），页面必须显式标注不可用 */
  hostMetricsAvailable: boolean
}

/**
 * 采集包含两次 CPU 采样，首次调用可能耗时数百毫秒，调用方需自行处理 loading。
 * silent：错误由页面内联展示，避免自动刷新时反复弹全局提示。
 */
export function fetchOverview() {
  return request<SystemOverview>({ url: '/system/overview', method: 'GET', silent: true })
}

File: internal/api/v1/handler/system.go
package handler

import (
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"github.com/novapanel/novapanel/internal/api/v1/response"
	"github.com/novapanel/novapanel/internal/repository"
	"github.com/novapanel/novapanel/internal/service/sysinfo"
)

// System 提供主机与面板运行信息，供总览页与运维页展示。
type System struct {
	collector *sysinfo.Collector
	db        *gorm.DB
}

// NewSystem 构造系统信息处理器。
func NewSystem(collector *sysinfo.Collector, db *gorm.DB) *System {
	return &System{collector: collector, db: db}
}

// OverviewResult 为总览接口响应。
type OverviewResult struct {
	sysinfo.Snapshot
	Database DatabaseStat `json:"database"`
}

// DatabaseStat 为数据库探活与连接池状态。
// 连接池细节只在受权限保护的总览接口里给，公开的 health 接口仍只回驱动与可用性。
type DatabaseStat struct {
	Driver          string `json:"driver"`
	OK              bool   `json:"ok"`
	Error           string `json:"error,omitempty"`
	OpenConnections int    `json:"openConnections"`
	InUse           int    `json:"inUse"`
	Idle            int    `json:"idle"`
}

// Overview 处理 GET /api/v1/system/overview。
// 采集失败的单项以零值返回，整体不报错：监控页不该因为一个字段读不到就整页空白。
func (h *System) Overview(c *gin.Context) {
	out := OverviewResult{Snapshot: h.collector.Collect()}
	out.Database = h.databaseCheck()
	response.OK(c, out)
}

// databaseCheck 与健康接口口径保持一致：Ping 通即视为可用。
func (h *System) databaseCheck() DatabaseStat {
	check := DatabaseStat{Driver: repository.Dialect(h.db), OK: true}
	sqlDB, err := h.db.DB()
	if err != nil {
		check.OK = false
		check.Error = err.Error()
		return check
	}
	if err := sqlDB.Ping(); err != nil {
		check.OK = false
		check.Error = err.Error()
		return check
	}
	stats := sqlDB.Stats()
	check.OpenConnections = stats.OpenConnections
	check.InUse = stats.InUse
	check.Idle = stats.Idle
	return check
}
